File: minesweeper.cpp
#include "minesweeper.hpp"

/* HELPER FUNCTIONS */

/** @brief 
* @return 0 for success, -1 for failure
*/
int Minesweeper::getMoveCoords(int& r, int&c) {
    std::string s;
    std::cout << "Input x: ";
    std::getline(std::cin,s);
    if (s.size() != 0 && isdigit(s)) {
        c = std::stoi(s); 
        if (c < 0 || c >= NUM_COLS) return -1;
    } else {
        return -1;
    }
    std::cout << "Input y: ";
    std::getline(std::cin,s); // take in y
    if (s.size() != 0 && isdigit(s)) {
        r = std::stoi(s);
        if (r < 0 || r >= NUM_ROWS) return -1;
    } else {
        return -1;
    }
    return 0;
}

/** @brief get number of bombs in 3x3 around (r,c) */
int Minesweeper::getNumBombs(int r, int c) {
    int count = 0;
    for (int i = -1; i <= 1; i++) {
        if (r+i >= 0 && r+i < NUM_ROWS) {
            for (int j = -1; j <= 1; j++) {
                if (c+j >= 0 && c+j < NUM_COLS && !(i==0 && j==0) && realBoard[r+i][c+j]) {
                    count += 1;
                }
            }
        }
    }
    return count;
}

/** @brief see if (r_test, c_test) is in 3x3 vicinity of (r,c) */
bool Minesweeper::inLocalThreeByThree(int r, int c, int r_test, int c_test) {
    for (int i = -1; i <=1; i++) {
        for (int j = -1; j <=1; j++) {
            if (r+i==r_test && c+j==c_test) {
                return true;
            }
        }
    }
    return false;
}

/** @brief reveal all the bombs, called by gameOver */
void Minesweeper::revealAllBombs() {
    for (int r = 0; r < NUM_ROWS; r++) {
        for (int c = 0; c < NUM_COLS; c++) {
            if (realBoard[r][c]) {
                gameBoard[r][c] = '*'; 
            }
        }
    }
}

/** @brief converts current gameBoard, realBoard to json file */
int Minesweeper::serializeGameState() {
    json gameState;
    gameState["numRows"] = NUM_ROWS;
    gameState["numCols"] = NUM_COLS;
    gameState["numBombs"] = NUM_BOMBS;
    
    gameState["gameBoard"] = json::array();
    for (int i = 0; i < NUM_ROWS; ++i) {
        json row = json::array();
        for (int j = 0; j < NUM_COLS; ++j) {
            row.push_back(gameBoard[i][j]);
        }
        gameState["gameBoard"].push_back(row);
    }

    gameState["realBoard"] = realBoard;
    gameState["bombWentOff"] = bombWentOff ? 1 : 0;
    gameState["numFlagsLeft"] = numFlagsLeft;
    gameState["numBombsFound"] = numBombsFound;

    std::ofstream outFile("assets/gameState.json");
    outFile << gameState.dump(4);
    
    return 0;
}

/** @brief loads json file into gameBoard, realBoard 
 * @return 0 for success, -1 if no gameState.json file to read
*/
int Minesweeper::unserializeGameState() {
    std::ifstream inFile("assets/gameState.json");
    if (!inFile.good()) return -1;
    json loadedState = json::parse(inFile);

    for (int r = 0; r < NUM_ROWS; r++) {
        for (int c = 0; c < NUM_COLS; c++) {
            gameBoard[r][c] = loadedState["gameBoard"][r][c].get<char>();
            realBoard[r][c] = loadedState["realBoard"][r][c].get<bool>();
        }
    }
    bombWentOff = (loadedState["bombWentOff"]==1) ? true : false;
    numFlagsLeft = loadedState["numFlagsLeft"];
    numBombsFound = loadedState["numBombsFound"];

    return 0;
}

/* CORE FUNCTIONS */

/** @brief displays board, number of flags used 
* @note takes r,c which is coord of current move, highlighted blue
**/
void Minesweeper::displayBoard(int current_r=-1, int current_c=-1) {
    // number of flags can be negative
    std::stringstream oss;
    oss << numFlagsLeft;
    std::cout << "Flags Left: " << oss.str() << "\n\n  ";
    for (int c = 0; c < NUM_COLS; c++) {
        std::cout << std::to_string(c).back() << " ";
    }
    std::cout << std::endl;

    for (int r = 0; r < NUM_ROWS; r++) {
        std::cout << std::to_string(r).back() << " ";
        for (int c = 0; c < NUM_COLS; c++) {
            if (current_r == r && current_c == c) { // current square, color blue
                std::cout << "\033[1;34m" << gameBoard[r][c] << "\033[0m ";
            } else if (gameBoard[r][c] == 'F') { // flag, color red
                std::cout << "\033[1;31mF\033[0m ";
            } else if (gameBoard[r][c] == '*') { // bomb, color magenta
                std::cout << "\033[1;35m*\033[0m "; 
            } else if (std::isdigit(gameBoard[r][c])) { // numbers, color green
                std::cout << "\033[1;32m" << gameBoard[r][c] << "\033[0m "; 
            } else { // nothing, or normal #
                std::cout << gameBoard[r][c] << " ";
            }
        }
        std::cout << std::endl;
    }
}

/** @brief reveal recursively around square */
void Minesweeper::reveal(int r, int c) {
    if (realBoard[r][c]) {
        gameBoard[r][c] = '*';
        bombWentOff = true;
    } else {
        int numBombs = getNumBombs(r,c);
        if (numBombs==0) {
            gameBoard[r][c] = ' ';
            for (int i = -1; i <= 1; i++) {
                if (r+i >= 0 && r+i < NUM_ROWS) {
                    for (int j = -1; j <=1; j++) {
                        if (c+j >= 0 && c+j<NUM_COLS && gameBoard[r+i][c+j] == '#' && !(i==0 && j==0)) {
                            reveal(r+i,c+j);
                        }
                    }
                }
            }
        } else {
            gameBoard[r][c] = ('0' + numBombs); 
        }
    }
}

/** @brief plays 1st step of game, generates board, saves board to gameState.json */
void Minesweeper::setupBoard() {
    // set up realBoard and gameBoard
    
    clearScreen();
    displayBoard();
    int r; 
    int c;

    int i = getMoveCoords(r,c);
    while (i==-1) {
        i = getMoveCoords(r,c);
    }

    for (int i = 0; i < NUM_BOMBS; i++) {
        int temp_r = randInt(0,NUM_ROWS-1);
        int temp_c = randInt(0,NUM_COLS-1);
        while (temp_r == r || temp_c == c || realBoard[temp_r][temp_c] || inLocalThreeByThree(r,c,temp_r,temp_c)) { 
            // keep finding new coords for bombs if there are already 
            // bombs there or in 3x3 around player 1st move
            temp_r = randInt(0,NUM_ROWS-1);
            temp_c = randInt(0,NUM_COLS-1);
        }
        realBoard[temp_r][temp_c] = true;
    }
    reveal(r,c);
    clearScreen();
    displayBoard();

    // should other players be able to see an empty board and maybe collab with player on which spot to pick for first move? - no for now
    serializeGameState();
}

/** @brief game over */
void Minesweeper::gameOver() {
    clearScreen();
    revealAllBombs();
    displayBoard();
    if(bombWentOff) {
        std::cout << "the minefield exploded..." << std::endl;
    } else {
        std::cout << "and so you survive another day..." << std::endl;
    }
}

/** @brief play one full iteration of the game */
void Minesweeper::playOneIteration() {
    int r;
    int c;
    int i = getMoveCoords(r,c);
    while (i == -1) {
        i = getMoveCoords(r,c);
    }
    // update screen board with blue highlight for selected coord
    clearScreen();
    displayBoard(r,c);
 
    std::cout << "[F]lag, [R]eveal, [U]nflag, or [Q]uit turn: ";
    
    std::string action;
    do {
        std::getline(std::cin, action);
    } while (action != "f" && action != "F" && action != "r" && action != "R" && action != "u" && action != "U" && action != "Q" && action != "q");
    
    // update boards
    if (action=="f" || action == "F") { // Flag
        if (gameBoard[r][c] == '#') {
            gameBoard[r][c] = 'F';
            numFlagsLeft--;
            if (realBoard[r][c]) numBombsFound++; // if actually a bomb
        }
    } else if (action == "r" || action == "R") { // Reveal
        if (gameBoard[r][c] == '#') {
            reveal(r,c);
        } 
    } else if (action == "u" || action == "U") { // unflag
        if (gameBoard[r][c] == 'F'){
            gameBoard[r][c] = '#';
            numFlagsLeft++;
        }
    } else { // was Q or q
        ;
    }
    clearScreen();
    displayBoard();
}

/** @brief networked version of playOneIteration */
void Minesweeper::attemptMove() {
    int r;
    int c;
    int i = getMoveCoords(r,c);
    while (i == -1) {
        i = getMoveCoords(r,c);
    }
    // update screen board with blue highlight for selected coord
    clearScreen();
    displayBoard(r,c);
 
    std::cout << "[F]lag, [R]eveal, [U]nflag, or [Q]uit turn: ";
    
    std::string action;
    do {
        std::getline(std::cin, action);
    } while (action != "f" && action != "F" && action != "r" && action != "R" && action != "u" && action != "U" && action != "Q" && action != "q");
    
    if ((action=="f" || action == "F") && gameBoard[r][c] == '#') { // Flag
        // send to server

    } else if ((action == "r" || action == "R") && gameBoard[r][c] == '#') { // Reveal
        // send to server
    } else if ((action == "u" || action == "U") && gameBoard[r][c] == 'F') { // unflag
    } else { // was Q or q
        ;
    }
    // only update board if affirmative response from server? (another gamestate.json file)
}

File: minesweeper.py
# minesweeper

# make sure color module is imported
import importlib.util
import sys

name = 'termcolor'

if name in sys.modules:
    print(f"{name!r} already in sys.modules")
elif (spec := importlib.util.find_spec(name)) is not None:
    # If you choose to perform the actual import ...
    module = importlib.util.module_from_spec(spec)
    sys.modules[name] = module
    spec.loader.exec_module(module)
else:
    print(f"can't find the {name!r} module")
    exit()

import random, re, os
from termcolor import colored

def main():
    print("Welcome to Minesweeper")
    
    global game_board
    global real_board # real board is a boolean 2D array, true for bomb
    global BOARD_SIZE; BOARD_SIZE = 10 

    game_board = [["#" for x in range(0,BOARD_SIZE)] for y in range(0,BOARD_SIZE)]
    real_board = [[False for x in range(0,BOARD_SIZE)] for y in range(0,BOARD_SIZE)]
    display_board(game_board)
    # F is marked/flag
    # space is nothing
    # 1-8 are the corresponding numbers
    
    global FLAGS; FLAGS = 13 # number of bombs
    global bomb; bomb = False
    coord_pattern = re.compile("^[0-9],[0-9]$")
    action_pattern = re.compile("^[F,R,U,f,r,u]$")
    action = ""
    flags_missing = FLAGS

    generate_board() # generate board, making sure not to generate a bomb on the square user chooses

    display_board(game_board)
    while(not (bomb or flags_missing==0)):
        next_coords = ""
        action = ""    
        while(not coord_pattern.match(next_coords)):
            next_coords = input("Input row,col: ")
        while(not action_pattern.match(action)):
            action = input("[F]lag, [R]eveal, or [U]nflag: ")
        
        coords = (int(next_coords.split(",")[0]),int(next_coords.split(",")[1]))
        if(game_board[coords[0]][coords[1]] == 'F' or game_board[coords[0]][coords[1]]=='#'):
            if(action.lower()=="f"):
                game_board[coords[0]][coords[1]] = colored('F','red')
                flags_missing -=1
            elif(action.lower()=="r"):
                reveal(coords) 
            elif(action.lower()=="u"):
                game_board[coords[0]][coords[1]] = "#"
                flags_missing +=1
            display_board(game_board)
    game_over()

def reveal(coords):
    if(real_board[coords[0]][coords[1]]):
        global bomb
        bomb =True
    else:    
        t = get_num_bombs(coords)
        if(t==0):
            game_board[coords[0]][coords[1]] = " "
            # cascade stuff
            for i in range(-1,2):
                if(coords[0]+i>=0 and coords[0]+i<BOARD_SIZE):
                    for j in range(-1,2):
                        if(coords[1]+j>=0 and coords[1]+j<BOARD_SIZE 
                        and game_board[coords[0]+i][coords[1]+j] == "#" and not(i==0 and j==0)):
                            reveal((coords[0]+i,coords[1]+j))
        else:
            game_board[coords[0]][coords[1]] = colored(t,'green')
    
def generate_board(): # the first move of the game
    # each 10x10 board has 13 bombs, makes sure start_coords is not occupied
    coord_pattern = re.compile("^[0-9],[0-9]$")
    start_coords = ""
    while(not coord_pattern.match(start_coords)):
        start_coords = input("Input row,col: ")
    start_coords = (int(start_coords.split(",")[0]),int(start_coords.split(",")[1]))
    
    bombs = set()

    # don't add bombs in 3x3 grid around where player first selects
    temp_coords = []
    for i in range(-1,2):
        if(start_coords[0]+i>=0 and start_coords[0]+i<BOARD_SIZE):
            for j in range(-1,2):
                if(start_coords[1]+j>=0 and start_coords[1]+j<BOARD_SIZE and not(i==0 and j==0)):
                    temp_coords.append((start_coords[0]+i,start_coords[1]+j))

    # actually add bombs
    while(len(bombs)<=FLAGS):
        temp = (random.randint(0,9),random.randint(0,9))
        while(temp==start_coords or temp in temp_coords): # make sure its not the starting coord or else it'd be annoying for user
            temp = (random.randint(0,9),random.randint(0,9))
        bombs.add(temp)
    for c in bombs:
        real_board[c[0]][c[1]] = True

    reveal(start_coords)

def get_num_bombs(coords): # get number of bombs around one tile
    count=0
    for i in range(-1,2):
        if(coords[0]+i>=0 and coords[0]+i<BOARD_SIZE):
            for j in range(-1,2):
                if(coords[1]+j>=0 and coords[1]+j<BOARD_SIZE and not(i==0 and j==0) and real_board[coords[0]+i][coords[1]+j]):
                    count+=1
    return count

def display_board(b):
    print("\n  ",end="")
    for i in range(BOARD_SIZE):
        print(str(i)+" ",end="")
    print("")
    for row in range(len(b)):
        print(str(row)+" ",end="")
        for col in range(len(b[0])):
            print(str(b[row][col])+" ", end="")
        print("")

def game_over(): # check if won/lost
    if(bomb):
        #os.system('cls' if os.name == 'nt' else 'clear')
        print(colored("the minefield exploded...",'blue'))
        exit()
    else:
        #os.system('cls' if os.name == 'nt' else 'clear')
        print(colored("and so you survive another day...",'blue'))
        exit()

if __name__ == "__main__":
    main()